### PR TITLE
Dịch "rnn-scratch_vn" - Phần 3

### LINKED ISSUE #1545: Dịch "rnn-scratch_vn" - Phần 3
Phần cần dịch: https://github.com/aivivn/d2l-vn/blame/master/chapter_recurrent-neural-networks/rnn-scratch_vn.md#L203-L279

### Hướng dẫn
- [Dịch trực tiếp trên website](https://github.com/aivivn/d2l-vn/blob/master/CONTRIBUTING.md).
- Sau khi nhấn vào phần cần dịch ở trên, hãy chọn **Normal view** và làm theo hướng dẫn ở trên để có thể bắt đầu dịch.

### PR DESCRIPTION
**Pull request by tranyenthy**
## Pull Request Checklist

**Đánh dấu tick (sau khi "Create Pull Request") vào các mục dưới đây:**

* [x] Pull Request này tương ứng với issue nào? Trả lời số issue sau dấu `#` (không có dấu cách): Close #1545 

* [x] Bản dịch này có bám sát **[Bảng thuật ngữ](https://github.com/aivivn/d2l-vn/blob/master/glossary.md)** không? Nếu một từ/cụm chưa có trong Bảng thuật ngữ, hãy tạo một Pull Request riêng đề xuất cách dịch từ/cụm đó.

* [x] Format của bản gốc có được giữ nguyên không?
    * Các phần in nghiêng, in đậm.
    * Tiêu đề (số lượng dấu `#` đầu dòng).
    * Bảng biểu, chú thích cho bảng (dòng phía trên mỗi bảng bắt đầu bằng dấu `:`).
    * Dịch các từ trong hình vẽ nếu cần. Các dòng có hình có dạng: `caption`.
    * Dịch các chú thích hình vẽ (thay các cụm `*dịch chú thích ảnh phía trên*` bằng bản dịch tiếng Việt).
    * Không dịch các phần code (nằm giữa hai dấu `````).
    * Copy các công thức toán từ bản gốc (các đoạn có `$`).
    * Giữ các dòng gán nhãn (bắt đầu với `:label:`).
    * Không tự thêm bớt các dòng trắng.
     
* [x] Trong một đoạn văn, mỗi câu văn đã được viết trong một dòng, giữa các dòng không có dòng trắng. Markdown sẽ coi những dòng liền nhau không có dòng trắng là một đoạn văn. Việc này giúp công đoạn review được thuận tiện hơn.

* [x] Tất cả chữ trong các hình vẽ đã được dịch. Nếu chưa, vui lòng xem thêm **[hướng dẫn hoặc yêu cầu dịch hình](https://github.com/aivivn/d2l-vn/blob/master/img/README.md)**.

* [x] Điền tên của bạn vào mục "Những người thực hiện" ở cuối file.

* [ ] **Sau khi được approve**, điền tên những reviewers mà bạn thấy có nhiều đóng góp cho Pull Request này. Xem danh sách Contributors **[tại đây](https://github.com/aivivn/d2l-vn/blob/master/docs/contributors_info.md)**.

## Với Reviewer

1. Chỉ review nếu Pull Request này đã sẵn sàng cho việc review.

2. Đã đọc kỹ hướng dẫn đóng góp và xem qua **[Bảng thuật ngữ](https://github.com/aivivn/d2l-vn/blob/master/glossary.md)**.

3. **Bạn có thể đề xuất thay đổi cách dịch mỗi dòng trực tiếp như trong hình dưới đây:**
img

### CODE REVIEW

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 1]
```diff
@@ -219,7 +219,10 @@ For the beginning of the sequence, we only update the hidden state.
 After that we begin generating new characters and emitting them.
 -->
 
-*dịch đoạn phía trên*
+Trước tiên chúng tôi giải thích chức năng dự đoán để có thể thường xuyên kiểm tra dự đoán trong quá trình huấn luyện.
+Hàm này dự đoán các ký tự `num_predicts` tiếp theo dựa trên `prefix` (một chuỗi chứa một vài ký tự).
+Để bắt đầu chuỗi, ta chỉ cập nhật trạng thái ẩn.
+Sau đó, ta bắt đầu tạo ra các ký tự mới và phát ra chúng.
```
Comment (quangnhat185) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:225]:
```suggestion
Trước tiên cần giải thích về hàm dự đoán để chúng ta có thể thường xuyên kiểm tra trong quá trình huấn luyện.
Hàm này dự đoán các ký tự `num_predicts` tiếp theo dựa trên `prefix` (một chuỗi chứa một vài ký tự).
Để bắt đầu chuỗi, ta cập nhật chỉ trạng thái ẩn.
Sau đó, ta bắt đầu tạo ra các ký tự mới và ban hành chúng.
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 2]
```diff
@@ -245,7 +248,9 @@ Given that we did not train the network, it will generate nonsensical prediction
 We initialize it with the sequence `traveller ` and have it generate 10 additional characters.
 -->
 
-*dịch đoạn phía trên*
+Trước tiên, ta kiểm tra hàm `predict_rnn`.
+Với việc ta đã không huấn luyện mạng, nó sẽ tạo ra các dự đoán vô nghĩa.
+Ta khởi tạo nó với chuỗi `traveller` và để nó tạo thêm 10 ký tự.
```
Comment (quangnhat185) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:253]:
```suggestion
Ta kiểm tra hàm `predict_rnn` trước tiên.
Vì mạng không được huấn luyện, nó sẽ tạo ra các dự đoán vô nghĩa.
Ta khởi tạo mạng với chuỗi `traveller` và để nó tạo thêm 10 ký tự.
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 3]
```diff
@@ -256,22 +261,25 @@ predict_ch8('time traveller ', 10, model, vocab, ctx)
 ## Gradient Clipping
 -->
 
-## *dịch tiêu đề phía trên*
+## Cắt bớt gradient
```
Comment (quangnhat185) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:264]:
```suggestion
## Cắt bớt Gradient
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 4]
```diff
@@ -256,22 +261,25 @@ predict_ch8('time traveller ', 10, model, vocab, ctx)
...
 <!--
 For a sequence of length $T$, we compute the gradients over these $T$ timesteps in an iteration, which results in a chain of matrix-products with length $\mathcal{O}(T)$ during backpropagating.
 As mentioned in :numref:`sec_numerical_stability`, it might result in numerical instability, e.g., the gradients may either explode or vanish, when $T$ is large.
 Therefore, RNN models often need extra help to stabilize the training.
 -->
 
-*dịch đoạn phía trên*
+Đối với chuỗi độ dài $T$, chúng tôi tính toán gradient theo các bước thời gian $T$ này trong một lần lặp, dẫn đến một dây chuyền các tích ma trận có độ dài $\mathcal{O}(T)$ trong quá trình lan truyền ngược.
+Như đã đề cập trong :numref:`sec_numerical_stability`, nó có thể dẫn đến mất ổn định số, ví dụ: các gradient có thể phát nổ hoặc biến mất, khi $T$ lớn.
+Do đó, các mô hình RNN thường cần thêm trợ giúp để làm ổn định việc huấn luyện.
```
Comment (quangnhat185) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:274]:
```suggestion
Đối với chuỗi có độ dài $T$, ta tính toán gradient theo các bước thời gian $T$ này trong một vòng lặp, dẫn đến một dây chuyền các tích ma trận có độ dài $\mathcal{O}(T)$ trong quá trình lan truyền ngược.
Như đã đề cập trong :numref:`sec_numerical_stability`, việc này có thể dẫn đến mất ổn định số, ví dụ: các gradient có thể phát nổ hoặc biến mất, khi $T$ lớn.
Do đó, các mô hình RNN thường cần thêm trợ giúp để ổn định việc huấn luyện.
```

File: chapter_recurrent-neural-networks/rnn-scratch_vn.md [hunk 5]
```diff
@@ -256,22 +261,25 @@ predict_ch8('time traveller ', 10, model, vocab, ctx)
...
 
 <!--
 Recall that when solving an optimization problem, we take update steps for the weights $\mathbf{w}$ in the general direction of the negative gradient $\mathbf{g}_t$ on a minibatch, 
 say $\mathbf{w} - \eta \cdot \mathbf{g}_t$. Let us further assume that the objective is well behaved, i.e., it is Lipschitz continuous with constant $L$, i.e.,
 -->
 
-*dịch đoạn phía trên*
+Hãy nhớ lại rằng khi giải quyết vấn đề tối ưu hóa, ta thực hiện các bước cập nhật cho các trọng số $\mathbf{w}$ theo hướng chung của gradient âm $\mathbf{g}_t$ trên một minibatch,
+ở đây là $\mathbf{w} - \eta \cdot \mathbf{g}_t$. Ta hãy giả định thêm rằng mục tiêu được xử lý tốt, tức là, đó là hàm Lipschitz liên tục với $L$, tức là,
```
Comment (quangnhat185) [inline chapter_recurrent-neural-networks/rnn-scratch_vn.md:282]:
```suggestion
Hãy nhớ lại rằng khi giải quyết vấn đề tối ưu hóa, ta thực hiện các bước cập nhật cho các trọng số $\mathbf{w}$ theo hướng chung của gradient âm $\mathbf{g}_t$ trên một minibatch,
ở đây là $\mathbf{w} - \eta \cdot \mathbf{g}_t$. Giả định rằng mục tiêu được xử lý tốt, khi đó hàm Lipschitz liên tục với biến $L$, tức là:
```

**Comment by duythanhvn** [comment]
@quangnhat185 Ông giáo để ý giúp mình mấy cái PR đang gắn label wip thì chưa vội review như này nha..

**Comment by duythanhvn** [comment]
Phần này khi nào có thể final nhỉ, cô giáo ơi?

**Comment by quangnhat185** [comment]
> @quangnhat185 Ông giáo để ý giúp mình mấy cái PR đang gắn label wip thì chưa vội review như này nha..

Hic my mistake. Cứ nhớ là bản draft có chữ [WIP] ở trước mà ko để ý status.

**Comment by tranyenthy** [comment]
> 
> 
> Phần này khi nào có thể final nhỉ, cô giáo ơi?

Trong cuối tuần mình lên nha @duythanhvn ơi.
